### PR TITLE
[SPARK-24215][PySpark][Follow Up] Implement eager evaluation for DataFrame APIs in PySpark

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

Address comments in #21370 and add more test.

## How was this patch tested?

Enhance test in pyspark/sql/test.py and DataFrameSuite

### CODE REVIEW

**Comment by xuanyuanking** [comment]
cc @HyukjinKwon @gatorsmile

**Comment by xuanyuanking** [comment]
retest this please

File: docs/configuration.md [hunk 1]
```diff
@@ -472,15 +472,17 @@ Apart from these, the following properties are also available, and may be useful
   <td>20</td>
   <td>
     Default number of rows in eager evaluation output HTML table generated by <code>_repr_html_</code> or plain text,
-    this only take effect when <code>spark.sql.repl.eagerEval.enabled</code> is set to true.
+    this only takes effect when <code>spark.sql.repl.eagerEval.enabled</code> is set to true. The valid range of this
```
Comment (gatorsmile) [inline docs/configuration.md]: For SQLConf, we do not need to hard code the conf description here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1314,6 +1314,31 @@ object SQLConf {
       "Other column values can be ignored during parsing even if they are malformed.")
     .booleanConf
     .createWithDefault(true)
+
+  val REPL_EAGER_EAVL_ENABLED = buildConf("spark.sql.repl.eagerEval.enabled")
+    .doc("Enable eager evaluation or not. If true and the REPL you are using supports eager " +
+      "evaluation, Dataset will be ran automatically. The HTML table which generated by " +
+      "<code>_repl_html_</code> called by notebooks like Jupyter will feedback the queries " +
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Can you rewrite these descriptions based on the description I posted in the original PR.

**Comment by gatorsmile** [comment]
Could you address the comments in the original PR?
- https://github.com/apache/spark/pull/21370#discussion_r194276735

File: docs/configuration.md (hunk 1 above)
Comment (xuanyuanking) [reply docs/configuration.md]: Got it, thanks, done in afada2b.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Thanks, done in afada2b.

**Comment by xuanyuanking** [comment]
```
Could you address the comments in the original PR?
```
Thanks, I want take this. Maybe it should be done in another jira and PR, and I should fix all the config hard code in PySpark？

File: python/pyspark/sql/conf.py [hunk 3]
```diff
@@ -64,6 +64,96 @@ def _checkType(self, obj, identifier):
                             (identifier, obj, type(obj).__name__))
 
 
+class ConfigEntry(object):
```
Comment (xuanyuanking) [inline python/pyspark/sql/conf.py]: @gatorsmile I moved all core configs using in pyspark into conf.py here. Please have a look when you have time.
https://github.com/apache/spark/pull/21370#discussion_r194276735

**Comment by xuanyuanking** [comment]
@gatorsmile I address the comments in the last commit, but maybe it should be done in a independent PR and Jira?

File: docs/configuration.md [hunk 4]
```diff
@@ -456,33 +456,6 @@ Apart from these, the following properties are also available, and may be useful
...
-  </td>
-</tr>
-<tr>
-  <td><code>spark.sql.repl.eagerEval.truncate</code></td>
-  <td>20</td>
-  <td>
-    Default number of truncate in eager evaluation output HTML table generated by <code>_repr_html_</code> or
-    plain text, this only take effect when <code>spark.sql.repl.eagerEval.enabled</code> set to true.
-  </td>
-</tr>
```
Comment (felixcheung) [inline docs/configuration.md]: we are removing documentation?

Comment (gatorsmile) [reply docs/configuration.md]: SQL Confs are not part of the documentation.

File: python/pyspark/sql/conf.py (hunk 3 above)
Comment (gatorsmile) [reply python/pyspark/sql/conf.py]: Thank you for fixing this! Let us do it in a separate PR.

File: python/pyspark/sql/conf.py [hunk 5]
```diff
@@ -64,6 +64,96 @@ def _checkType(self, obj, identifier):
...
+        """Give a default value for current config entry, the default value will be set
+        to _NoValue when its absent"""
+        self.default = default
+        return self
+
+    def read(self, ctx):
+        """Read value from this config entry through sql context"""
+        return self.converter(ctx.getConf(self.confKey, self.default))
+
+class SQLConf(object):
```
Comment (viirya) [inline python/pyspark/sql/conf.py]: I think this PySpark SQLConf stuff should be done in a separate Jira/PR.

Comment (HyukjinKwon) [reply python/pyspark/sql/conf.py]: Yea, it should be separate.

File: python/pyspark/sql/conf.py [hunk 6]
```diff
@@ -64,6 +64,96 @@ def _checkType(self, obj, identifier):
...
+        .withDefault("20")
+
+    PANDAS_RESPECT_SESSION_LOCAL_TIMEZONE = \
+        ConfigEntry("spark.sql.execution.pandas.respectSessionTimeZone")\
+        .boolConf()
+
+    SESSION_LOCAL_TIMEZONE = ConfigEntry("spark.sql.session.timeZone")\
+        .stringConf()
+
+    ARROW_EXECUTION_ENABLED = ConfigEntry("spark.sql.execution.arrow.enabled")\
```
Comment (HyukjinKwon) [inline python/pyspark/sql/conf.py]: This duplicates the key. I think current way duplicates a lot of codes in Scala side.

File: python/pyspark/sql/conf.py (hunk 3 above)
Comment (xuanyuanking) [reply python/pyspark/sql/conf.py]: Yep, done in #21648

File: python/pyspark/sql/conf.py (hunk 5 above)
Comment (xuanyuanking) [reply python/pyspark/sql/conf.py]: Yeah, agree, done in #21648.

File: python/pyspark/sql/conf.py (hunk 6 above)
Comment (xuanyuanking) [reply python/pyspark/sql/conf.py]: Yep, I'm also puzzled by this, cause we also do the register in Scala side. How about just call buildConf on Scala side for theses keys which used only on PySpark? Lets discuss it in #21648

File: docs/configuration.md (hunk 4 above)
Comment (felixcheung) [reply docs/configuration.md]: this should be in `sql-programming-guide.md` right?

**Comment by xuanyuanking** [comment]
In the last commit I revert the changes of SQLConf and created a new PR of #21648. Could this follow up PR merged first? Thanks.

File: docs/configuration.md (hunk 4 above)
Comment (xuanyuanking) [reply docs/configuration.md]: Follow the SQL configuration, all the description can be shown by `spark.sql("SET -v").show(numRows = 200, truncate = false)`. https://spark.apache.org/docs/latest/configuration.html#spark-sql

**Comment by xuanyuanking** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM Thanks! Merged to master.

**Pull request closed without merging**